### ISSUE TITLE
Workspace switcher popup doesn't close automatically on Gnome 42.1 on Wayland

### ISSUE THREAD

**Issue by tstirrat15** [mzur/gnome-shell-wsmatrix#227]
## Description
I use `ctrl + alt + <arrow>` to swap between workspaces. In previous versions of gnome, when I released those keys, the overlay that shows which workspace is focused would close itself and I could return to whatever I was doing on the target screen. Currently, after I press the key combo, the switcher stays on screen, and it catches all keyboard events until I hit `esc`.

It's workable, because all I have to do is hit `esc`, but it's additional friction and is frustrating.

Please let me know if there's additional context that I can provide.

**Comment by mzur** [comment]
Sounds like a duplicate of #219. Which GNOME version do you use?

**Comment by ebeem** [comment]
I would like to also know if you're using `xorg` or `wayland`
I think it's still buggy under `wayland` for some reason.

Of course this is still an upstream issue.

**Comment by reify-tanner-stirrat** [comment]
Yep, it's a duplicate of #219 and seems to be because ubuntu 22.04 hasn't taken gnome 42.1 yet. I'm still on 42. Thank you for pointing me in that direction!

**Comment by tstirrat15** [comment]
I've been using Wayland, which apparently has problems with the most recent Nvidia drivers as well. Sounds like this is a bleeding-edge tax kind of thing.

**Comment by ElijahLynn** [comment]
I'm on 42.1 and just went from X11 to Wayland and still have this issue, fwiw.

**Comment by ElijahLynn** [comment]
> I'm reopening this as I'm seeing the same.

Thanks for validating. Should we update the title to include that it is Gnome 42 on Wayland?

**Comment by tstirrat15** [comment]
Yep, done.

**Comment by tstirrat15** [comment]
I just received version `42.1-0ubuntu1` of `mutter` in an apt upgrade and now it's behaving as desired. Closing again.

**Comment by mzur** [comment]
Can confirm. It wasn't GNOME Shell 42.1 that fixed this issue but mutter 42.1.

**Comment by dseomn** [comment]
I'm seeing this issue again. Debian testing, mutter 42.2-1, gnome-shell 42.2-1.

**Comment by mzur** [comment]
We have GNOME Shell 42.4 now (on Ubuntu). Does anyone still see this behavior with the newest version?

**Comment by dseomn** [comment]
I'm still seeing this with mutter 42.4-2 and gnome-shell 42.4-1+build1.

**Comment by mzur** [comment]
Does anyone see this in GNOME Shell 45?

**Comment by tstirrat15** [comment]
What I'm experiencing in v45 is that the tiling works (I have a 2x2 setup), moving between them with the arrow keys works, but there's some animation weirdness where if the layout is:

```
1 2
3 4
```

moving vertically from 1-3 or 2-4 is animated as a horizontal swipe. It's a little visually jarring but not broken.

**Comment by mzur** [comment]
The movement animation is one of the things that won't work until GNOME Shell 45.1, so that's expected :wink:

**Comment by dseomn** [comment]
> Does anyone see this in GNOME Shell 45?

Sorry for the late reply, Debian just upgraded gnome recently. I'm still seeing the same issue with mutter 46.3.1-2 and gnome-shell 46.3.1-2.

**Comment by dseomn** [comment]
I just tried creating a new user account and setting up my dotfiles in it, and the issue was gone there. But it's still an issue in my existing user account. I compared the running processes of both users and didn't see anything relevant. I reset a bunch of dconf keys in the old user account.

At this point I really don't understand what's going on, but if nobody else is still seeing the issue it's probably safe to close it, since I can't reproduce it with a fresh user account anymore.

**Comment by mzur** [comment]
Do you have the same GNOME extensions in both accounts? Maybe there are conflicts between extensions. Or you use X11 in one and Wayland in the other?

**Comment by dseomn** [comment]
> Do you have the same GNOME extensions in both accounts? Maybe there are conflicts between extensions.

My dotfiles [configure that](https://github.com/dseomn/dotfiles/blob/aa681ad3492ab9bd2980c75ea4d6333c75f99247/.config/dconf-load-yaml/conf.d/gnome-shell-00.yaml#L25-L34), so it should be the same.

> Or you use X11 in one and Wayland in the other?

Wayland in both, just double checked.

I also just compared the output of `env` between the two users, and didn't see any significant differences. (A bunch of variables that have the username or user ID differ in that part, some PIDs are different, some random characters in tmp paths are different, DISPLAY has a different number, but nothing significant.)

**Comment by dseomn** [comment]
After a bunch more digging, I've got two similar but different dconf database files, one of which causes the bug on wayland and one that doesn't. Here's the diff between the two. I think the `org/gnome/Ptyxis` part is just because I opened a terminal at some point to run dconf commands, and I'm guessing `org/gnome/software` is because of some notifications. So I'm focusing on `numlock-state` now.

Using the dconf DB with the bug, I reset `numlock-state`, logged out, logged back in again. Not only was the bug still present, but `numlock-state` was set to true again. Maybe there's some software running in the session that writes to `numlock-state` after I ran `dconf reset`? Also, I've been doing this testing on a laptop that doesn't have a numlock key or indicator on its keyboard.

I think the next step is to try to reset `numlock-state` while gnome isn't running and see if that fixes the bug.

```
dseomn@helicon:~$ diff -u <(DCONF_PROFILE=<(echo user-db:user-without-bug) dconf dump /) <(DCONF_PROFILE=<(echo user-db:user-with-bug-2) dconf dump /)
--- /dev/fd/63  2025-02-03 17:23:19.426279592 -0500
+++ /dev/fd/62  2025-02-03 17:23:19.426279592 -0500
@@ -1,3 +1,8 @@
+[org/gnome/Ptyxis]
+default-profile-uuid='97e2892d4eb145776cd3bd1f67a13f2b'
+profile-uuids=['97e2892d4eb145776cd3bd1f67a13f2b']
+window-size=(uint32 211, uint32 54)
+
 [org/gnome/desktop/app-folders]
 folder-children=['Utilities', 'YaST', 'Pardus']
 
@@ -26,6 +31,9 @@
 [org/gnome/desktop/notifications/application/org-gnome-software]
 application-id='org.gnome.Software.desktop'
 
+[org/gnome/desktop/peripherals/keyboard]
+numlock-state=true
+
 [org/gnome/desktop/wm/preferences]
 num-workspaces=4
 
@@ -49,9 +57,10 @@
 locations=@av []
 
 [org/gnome/software]
+check-timestamp=int64 1738619810
 first-run=false
 install-timestamp=int64 1701989901
-update-notification-timestamp=int64 1738620128
+update-notification-timestamp=int64 1738619880
 
 [org/gtk/gtk4/settings/file-chooser]
 show-hidden=false
```

**Comment by dseomn** [comment]
It's `numlock-state`!

I was able to modify the dconf DB outside of a graphical session by running `dbus-launch -- dconf ...` instead of `dconf ...`. With `numlock-state` unset, the popup closes normally. With `numlock-state` set to true, it doesn't close until I press Escape.

Then I plugged in an external keyboard with a numlock key and light. Pressing the numlock key seems to both toggle the value of `numlock-state` in dconf *and* affect the popup hiding bug. Writing directly to dconf without using the keyboard key toggles the value in dconf but does not affect the popup bug. So as far as I can tell, the behavior is affected only by the initial state of `/org/gnome/desktop/peripherals/keyboard/numlock-state` when the session starts and by subsequent presses of a physical numlock button, but not by direct dconf changes.

@mzur Any chance you were testing on a keyboard without numlock, or with numlock off? Are you able to reproduce the issue with numlock on?

**Comment by mzur** [comment]
Can confirm. With numlock on the popup doesn't disappear. Thanks for digging into this! Do you think this is something that can/should be fixed in this extension or is this an upstream issue?

**Comment by dseomn** [comment]
> Do you think this is something that can/should be fixed in this extension or is this an upstream issue?

Not sure. I started looking into that too yesterday, but I don't really understand the problem well enough yet. Here's a summary of what I'm thinking though in case it helps.

I think this is the relevant code in wsmatrix, but you'd probably know that better than me: https://github.com/mzur/gnome-shell-wsmatrix/blob/a8aed7e8b4641cbd09e217211e11b5896381acaf/wsmatrix%40martin.zurowietz.de/workspacePopup/workspaceManagerOverride.js#L469-L471

I think https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/a73a4f8455b1374b38724a532137a9b614a7fb19/js/ui/switcherPopup.js#L133-146 and https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/a73a4f8455b1374b38724a532137a9b614a7fb19/js/ui/switcherPopup.js#L213-218 are the relevant parts of gnome-shell, and I'm *guessing* the second one is more relevant here. https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/a73a4f8455b1374b38724a532137a9b614a7fb19/js/ui/switcherPopup.js#L25-35 might be relevant too, not sure.

https://gitlab.gnome.org/GNOME/mutter/-/blob/c250f602bdc58b5614afc67baf96ea45f5bca9c6/clutter/clutter/clutter-enums.h#L239 seems to be where `Clutter.ModifierType` is defined, and notably, `CLUTTER_MODIFIER_MASK` (which I'm guessing is the same as `Clutter.ModifierType.MODIFIER_MASK` in the JS code?) includes both `CLUTTER_LOCK_MASK`[^1] and `CLUTTER_MOD2_MASK`[^2].

I don't know if https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/03e5d4bdd89b695e4699862cd8efe61a907aa37c/js/ui/popupMenu.js?page=2#L1070-1074 is related or not, but the way it clears the caps and num lock modifiers is potentially interesting at least.

So I'm not sure what the solution is yet, but I'm leaning towards thinking that mod2 and possibly lock masks should be cleared from some variables, somewhere. Maybe in this extension, maybe in gnome-shell.

[^1]: I haven't tested the bug with caps lock on yet, but I think there's a decent chance that caps lock also causes the bug.
[^2]: https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/blob/59cc11f6d405010aaa1fa9585cab2f261a473398/symbols/pc#L33 shows that numlock is mod2, I think. I'm a bit rusty at reading xkb data files though.

**Comment by dseomn** [comment]
Does anybody know an easy way to test changes to extension code in wayland without logging out and logging back in again? I tried `dbus-run-session -- gnome-shell --nested --wayland` but I can't figure out how to get it to capture key events.

**Comment by dseomn** [comment]
TL;DR: I think I do understand the bug now. And fixing it is as easy as clearing the caps and num lock bits, but I don't know whether to do that here or in gnome-shell.

I added a `console.log` statement after this line to see what it was returning: https://github.com/mzur/gnome-shell-wsmatrix/blob/a8aed7e8b4641cbd09e217211e11b5896381acaf/wsmatrix%40martin.zurowietz.de/workspacePopup/workspaceManagerOverride.js#L470

On Xorg, it logged 0b1100 regardless of the caps or num lock state.

On Wayland, it logged 0b1100 when caps and num lock were off, 0b1110 with caps lock, 0b11100 with numlock, and 0b11110 with both.

I was using Ctrl+Alt and arrow keys to trigger the switcher, and https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/blob/59cc11f6d405010aaa1fa9585cab2f261a473398/symbols/pc#L32 shows Mod1 as the modifier for the Alt keys. So going back to https://gitlab.gnome.org/GNOME/mutter/-/blob/c250f602bdc58b5614afc67baf96ea45f5bca9c6/clutter/clutter/clutter-enums.h#L239 0b1100 matches Ctrl+Alt. And the additional bits in Wayland match the modifiers for caps and num lock.

https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/a73a4f8455b1374b38724a532137a9b614a7fb19/js/ui/switcherPopup.js#L25-35 seems to find the highest bit, and from some testing it seems that only that bit matters for the popup hiding. Caps lock doesn't trigger the bug for me because it's lower than both Ctrl and Alt. But if somebody had a keyboard shortcut to switch workspaces using only Shift, I bet caps lock would trigger the bug. Num lock is higher than both Ctrl and Alt, so it triggers the bug for me.

Is it a bug that Xorg doesn't pass the modifiers for caps and num lock? Or is it a bug that Wayland does? I'm not sure, but I'd lean towards thinking it's a bug in Xorg to not pass all the modifiers.

I don't see any documentation for https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/a73a4f8455b1374b38724a532137a9b614a7fb19/js/ui/switcherPopup.js#L106 so I also don't know if the `mask` parameter is *supposed* to be a mask of all modifiers (including caps and num lock) or just the modifiers that triggered the popup (excluding the two locks presumably, but who knows, maybe somebody uses those in keyboard shortcuts?). So if the Wayland behavior of `event.get_state()` is correct and wsmatrix should have any visibility into caps and num lock, then I don't know if the bug is in wsmatrix for passing those to `SwitcherPopup.show` or in `SwitcherPopup.show` for not filtering those out.

I tested clearing the num and caps lock bits in wsmatrix, and it did fix the bug for me.

@mzur do you think it makes sense to reach out to gnome-shell folks (by filing a bug?) about where caps and num lock should be filtered out? Or should I just make a PR in this repo?

**Comment by dseomn** [comment]
> I don't know if the bug is in wsmatrix for passing those to SwitcherPopup.show or in SwitcherPopup.show for not filtering those out.

I tried to figure it out by looking at [gnome-shell's own code that uses SwitcherPopup](https://gitlab.gnome.org/search?group_id=8&nav_source=navbar&project_id=546&repository_ref=main&scope=blobs&search=extension%3Ajs+SwitcherPopup&search_code=true). All the calls I could find passed something like `binding.get_mask()` instead of something like `event.get_state()`. I didn't trace all the `binding` variables to be sure, but that does look to me like `SwitcherPopup.show` is designed to take a modifier mask from configured key bindings, which generally would not include caps or num lock.

I also tested the Ctrl+Alt+Tab switcher in gnome-shell on wayland, since I think that uses SwitcherPopup. It didn't have the hiding bug, which goes along with my theory above.

I'll start working on a PR to this repo.

**Comment by mzur** [comment]
@dseomn Awesome, thanks a lot! I think fixing this in the extension is the quickest way. I'll review/test/merge your PR as soon as I can!

**Comment by cohnt** [comment]
~~I'm experiencing this issue on 24.04 with gnome 46.2 and extension version 50 -- it just started after I updated mutter. (This is with timeout 0, so maybe related to #227? And since it's gnome 46.2, maybe #279?)~~

~~Notably, Super+Alt+Arrow (instead of Ctrl+Alt+Arrow) doesn't have this issue, and the popup closes as expected. But only left and right work, and they just swap between the top-left and bottom-right workspaces. And I'd prefer to not change by keyboard shortcuts if posssible.~~

EDIT: And now the issue has randomly vanished. Looks like it was associated with my desktop randomly switching to Wayland for a bit. It seems to have set itself back to X11, and all is well again.
